Задание спринта №4 выполнено.

ArtMarenkov opened the pull request.

ArtMarenkov committed:
Задание спринта №4 выполнено.

kadirovgm commented on pages/mainpage.py:
```diff
@@ -0,0 +1,40 @@
+class MainPage:
+
+
+    exp_t_1 = 'Сутки — 400 рублей. Оплата курьеру — наличными или картой.'
```
супер, что вынес в переменные эти статичные тестовые данные.

Можно лучше: тестовые статичные данные лучше выносить в отдельный файл типа [data.py](http://data.py/) или [urls.py](http://urls.py/) 
Это нужно для легкой поддержке проекта и переиспользовании тестовых данных, мы должны стремиться к тому, чтобы разделять код и тестовые данные, тем самым избегая повторений.

kadirovgm commented on pages/mainpage.py:
```diff
@@ -0,0 +1,40 @@
+from selenium.webdriver.support import expected_conditions
+from selenium.webdriver.support.wait import WebDriverWait
+import allure
+class MainPage:
```
Нужно исправить: PEP8 надо поправить структуру кода (оставить пустые строки после импортов и не оставлять внутри класса)

ArtMarenkov replied:
Исправлено.

kadirovgm commented on pages/mainpage.py:
```diff
@@ -0,0 +1,40 @@
+    question_text = 'accordion__panel-'
+
+    @allure.step('Открываем страницу в браузере')
+    def __init__(self, driver):
```
нужно исправить: нужно создать базовый класс Base Page и инициализацию браузера вынести туда, а в POM-файлах/классах наследоваться от BasePage

класс BasePage - это такой родительский класс, в котором содержатся общие методы, такие как: клик по определенному элементу, скролл к определенному элементу, ожидание определенного элемента, изменение табов итд.

То есть эти методы не привязаны к определенной странице, и могут быть переиспользованы в любой POM-странице, стоит сюда добавить именно такие методы.

ArtMarenkov replied:
Исправлено.

kadirovgm commented on pages/mainpage.py:
```diff
@@ -0,0 +1,40 @@
+from selenium.webdriver.support import expected_conditions
+from selenium.webdriver.support.wait import WebDriverWait
+import allure
+class MainPage:
```
нужно исправить: нужно наследоваться от базового класса Base Page (ниже оставил коммент о нем)

ArtMarenkov replied:
Исправлено.

kadirovgm commented on pages/mainpage.py:
```diff
@@ -0,0 +1,40 @@
+        self.driver.get('https://qa-scooter.praktikum-services.ru/')
+
+    @allure.step('Прокручиваем страницу вниз.')
+    def scroll_page_down(self):
```
нужно исправить: если хочешь оставить методом главной страницы то конкретизируй метод, чтобы в названии было точно указано, до куда хочешь проскроллить

но я бы советовал скролл к определенному элементу (передав локатор в аргуементы) вынести как общий метод класса BasePage, ведь скролл к определенному элементу можно использовать во всех POM-страницах

ArtMarenkov replied:
Исправлено.

kadirovgm commented on pages/mainpage.py:
```diff
@@ -0,0 +1,40 @@
+
+    @allure.step('Прокручиваем страницу вниз.')
+    def scroll_page_down(self):
+        q_first = self.driver.find_element(By.ID, 'accordion__heading-0')
```
нужно исправить: локаторы не держим в методах или тестах, выносим в отедльный класс и файл, чтобы была единая точка обращения к локаторам, таким образом мы улучшим поддержку кейсов и переиспользуемость локаторов

* коммент актуален по всему коду

ArtMarenkov replied:
Исправлено.

kadirovgm commented on pages/mainpage.py:
```diff
@@ -0,0 +1,40 @@
+        self.driver.execute_script('arguments[0].scrollIntoView();', q_first)
+
+    @allure.step('Выбираем вопрос № {param}')
+    def click_on_question(self, param):
```
супер, хороший метод главной страницы!

kadirovgm commented on pages/orderpage.py:
```diff
@@ -0,0 +1,188 @@
+
+class OrderPage:
+
+    up_order_button = "[class *= 'Header_Nav'] [class *= 'Button']"
```
нужно исправить: давай попробуем оптимизировать этот локатор, чтобы он выглядел стабильнее, лаконичнее, кратко и читаемо

ArtMarenkov replied:
@kadirovgm, Добрый вечер.
А какой вариант локатора предпочтительнее? 
1) [class = 'Button_Button__ra12g'] - постфикс является автогенерённым и может быть изменён.
2) Если использовать локатор через text() = 'Заказать', то при изменении текста кнопки тоже локатор перестанет работать.
3) Или какой-то другой подход имелся в виду?

Заранее благодарен.

ArtMarenkov replied:
По согласованию с куратором (Алена Тямейкина) указанные локаторы исправил через использование xpath и text()

ArtMarenkov replied:
Исправлено.

kadirovgm commented on pages/orderpage.py:
```diff
@@ -0,0 +1,188 @@
+    colour_black = "[id = 'black']"
+    colour_grey = "[id = 'grey']"
+    comment = "[placeholder = 'Комментарий для курьера']"
+    make_an_order_button = "[class *= 'Button'] [class = 'Button_Button__ra12g Button_Middle__1CSJM']"
```
нужно исправить: давай попробуем оптимизировать этот локатор, чтобы он выглядел стабильнее, лаконичнее, кратко и читаемо

ArtMarenkov replied:
Исправлено.

kadirovgm commented on pages/orderpage.py:
```diff
@@ -0,0 +1,188 @@
+    colour_grey = "[id = 'grey']"
+    comment = "[placeholder = 'Комментарий для курьера']"
+    make_an_order_button = "[class *= 'Button'] [class = 'Button_Button__ra12g Button_Middle__1CSJM']"
+    order_confirmation_button = "[class = 'Order_Modal__YZ-d3'] [class = 'Order_Buttons__1xGrp'] [class = 'Button_Button__ra12g Button_Middle__1CSJM']"
```
нужно исправить: давай попробуем оптимизировать этот локатор, чтобы он выглядел стабильнее, лаконичнее, кратко и читаемо

ArtMarenkov replied:
Исправлено.

kadirovgm commented on pages/orderpage.py:
```diff
@@ -0,0 +1,188 @@
+    dzen_main_page = "[class *= 'dzen']"
+
+    @allure.step('Открываем страницу в браузере')
+    def __init__(self, driver_fixt):
```
нужно исправить: поместить в Base Pаge, при наследовании init также наследуется

ArtMarenkov replied:
Исправлено.

kadirovgm commented on pages/orderpage.py:
```diff
@@ -0,0 +1,188 @@
+from selenium.webdriver.support.wait import WebDriverWait
+import allure
+
+class OrderPage:
```
Нужно исправить: Если придерживаться классических рекомендации POM, то тебе нужно наследоваться в этом классе от BasePage - как раз таки для того, чтобы наследовать его общие методы

ArtMarenkov replied:
Исправлено.

kadirovgm commented on pages/orderpage.py:
```diff
@@ -0,0 +1,188 @@
+
+    @allure.step('Прокручиваем страницу вниз.')
+    def scroll_page_down(self):
+        # q_first = self.driver.find_element(By.ID, 'accordion__heading-0')
```
нужно исправить: не мержим закомменченный код в общую ветку

ArtMarenkov replied:
Исправлено.

kadirovgm commented on pages/orderpage.py:
```diff
@@ -0,0 +1,188 @@
+        self.driver.find_element(By.CSS_SELECTOR, self.up_order_button).click()
+
+    @allure.step('Прокручиваем страницу вниз.')
+    def scroll_page_down(self):
```
нужно исправить: как видишь, этот метод снова используешь и прописываешь в другом классе, поэтому стоит вынести как метод base page и переиспользовать его в POM файлах

ArtMarenkov replied:
Исправлено.

kadirovgm commented on pages/orderpage.py:
```diff
@@ -0,0 +1,188 @@
+    @allure.step('Заполняем поле "Фамилия".')
+    def set_surname(self):
+        WebDriverWait(self.driver, 3).until(expected_conditions.element_to_be_clickable((By.CSS_SELECTOR, self.surname_field)))
+        self.driver.find_element(By.CSS_SELECTOR, self.surname_field).click()
```
можно лучше: проверь пожалуйста, обязательно ли кликать перед тем как вставить текст в поле

ArtMarenkov replied:
Исправлено.

kadirovgm commented on pages/orderpage.py:
```diff
@@ -0,0 +1,188 @@
+
+        return self.driver.find_element(By.CSS_SELECTOR, self.dzen_main_page).is_displayed()
+
+    def make_order(self, button_place, colour):
```
супер! молодец, что объединил для краткости шагов тест-кейса

kadirovgm commented on test/test_main_page.py:
```diff
@@ -0,0 +1,25 @@
+    driver = None
+
+    @classmethod
+    def setup_class(cls):
```
нужно исправить: у тебя в conftest.py уже есть фикстура, выполняющая роль открытия и закрытия браузера для каждого тестового метода - это правильно.

Поэтому setup и teardown методы не нужно отдельно прописывать в тестах

ArtMarenkov replied:
Исправлено.

kadirovgm commented on test/test_main_page.py:
```diff
@@ -0,0 +1,25 @@
+        assert exp_text == main_page.get_q_text(param)
+
+    @classmethod
+    def teardown_class(cls):
```
нужно исправить: нужно убрать в пользу фикстур, оставил коммент выше

ArtMarenkov replied:
Исправлено.

kadirovgm commented on test/test_main_page.py:
```diff
@@ -0,0 +1,25 @@
+        main_page.scroll_page_down()
+        main_page.click_on_question(param)
+
+        assert exp_text == main_page.get_q_text(param)
```
можно лучше: для лучшего отчета можно добавить assert message

ArtMarenkov replied:
Исправлено.

kadirovgm commented on test/test_order_page.py:
```diff
@@ -0,0 +1,21 @@
+
+        order_page = OrderPage(driver_fixt)
+        order_page.make_order(button_place, colour)
+
```
нужно исправить: придерживаемся правила атомарности тестов, один тест-одна проверка (один ассерт), так что тут лучше для каждого сценария новый тест написать

ArtMarenkov replied:
Исправлено.

ArtMarenkov committed:
Внесены испрвления по замечаниям к проекту.
1) Создана директория 'data' с файлами:
data.py - содержит перечень ожидаемых ответов на вопросы для главной страницы в автотесте из test\test_main_page.py;
locators.py - содержит все локаторы, используемые в методах.
2) В директорию 'pages' добавлен файл base_page.py (базовый класс), в который перенесены общие методы для всех страниц и инициализация драйвера.
Файлы mainpage.py и orderpage.py переименованы в main_page.py и order_page.py соответственно.
3) Внутри классов убраны пустые строки и добавлены пустые строки между импортов и классом.
4) Оптимизированы некоторые локаторы.
5) Убраные закомментированные строки.
6) Оптимизированы методы для ввода тексовых значений в поля на странице (убраны предварительные клики по полю).
7) Из test_main_page.py убраны методы setup и teardown.
8) В test_main_page.py и order_page.py добавлены assert message.
9) В order_page.py переписаны тесты в соответствии с правилом атомарности тестов.

kadirovgm commented on .gitignore:
```diff
@@ -0,0 +1,241 @@
+### JetBrains template
+# Covers JetBrains IDEs: IntelliJ, RubyMine, PhpStorm, AppCode, PyCharm, CLion, Android Studio, WebStorm and Rider
```
Можно лучше: в файле гитигнора лучше оставлять только те локальные файлы, что не должны быть запушены в общий репозиторий, в твоем случае ты использовала общий шаблон, это не совсем верно, мы должны избегать использования ненужного кода в проекте.

kadirovgm commented on conftest.py:
```diff
@@ -0,0 +1,8 @@
+
+@pytest.fixture
+def driver_fixt():
+    driver = webdriver.Firefox()
```
супер! но можно сюда еще добавить открытие главной страницы сайта, чтобы по умолчанию открывалась для каждого теста

kadirovgm commented on data/data.py:
```diff
@@ -0,0 +1,9 @@
+class Expected_data():
```
отлично!

kadirovgm commented on data/locators.py:
```diff
@@ -0,0 +1,25 @@
+class Locators:
```
- Можно лучше: класс с локаторами лучше разделить на несколько: например `AuthPageLocators`, `MainPageLocators`
Сейчас это не критично, но в будущем это точно стоит сделать, так как таким образом ты не будешь импортить лишние локаторы в файл с тестами на экран авторизации, например. Т.е. у каждой страницы должен быть свой класс с локаторами.
    
    При таком подходе код станет лаконичнее, легко поддерживаемой. Ведь у экрана авторизации и у экрана главной могут быть свои NEXT_BUTTON например. Тогда можно будет легко оба обозвать именно так:
    `MainPageLocators.NEXT_BUTTON` и `AuthPageLocators.NEXT_BUTTON`

kadirovgm commented on pages/base_page.py:
```diff
@@ -0,0 +1,44 @@
+class Base_page():
+    def __init__(self, driver_fixt):
+        self.driver = driver_fixt
+        driver_fixt.get("https://qa-scooter.praktikum-services.ru/")
```
нужно исправить: лучше это вынести в фикстуру driver

kadirovgm commented on pages/base_page.py:
```diff
@@ -0,0 +1,44 @@
+
+class Base_page():
+    def __init__(self, driver_fixt):
+        self.driver = driver_fixt
```
можно лучше: ничего страшного если даже написать `self.driver = driver`

kadirovgm commented on pages/base_page.py:
```diff
@@ -0,0 +1,44 @@
+
+    def check_page_url(self, url):
+        WebDriverWait(self.driver, 10).until(lambda driver: self.driver.current_url == url)
+
```
супер! отличные базовые методы, молодец

kadirovgm commented on pages/order_page.py:
```diff
@@ -0,0 +1,138 @@
+
+    @allure.step('Нажимаем на верхнюю кнопку "Заказать".')
+    def go_to_order_page_up(self):
+        Base_page.wait_for_element_is_clickable(self, By.XPATH, Locators.up_order_button)
```
можно лучше: так как ты наследовал все методы BasePage, то здесь на его методы обращаться просто через self

self.wait_for_element_is_clickable()

kadirovgm commented on pages/order_page.py:
```diff
@@ -0,0 +1,138 @@
+        Base_page.check_page_url(self, "https://dzen.ru/?yredirect=true")
+        return Base_page.is_element_displayed(self, By.CSS_SELECTOR, Locators.dzen_main_page)
+
+    def make_order(self, button_place, colour):
```
супер!
